### PR TITLE
Distance Queries in Tree Editorial

### PR DESCRIPTION
**Pull request by TheGamingMousse**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

### CODE REVIEW

File: solutions/gold/ac-distancequeries.mdx [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+// BeginCodeSnip{Binary Indexed Tree (Click to expand)}
```
Comment (SansPapyrus683) [inline solutions/gold/ac-distancequeries.mdx:51]: just have it say "BIT (from the module)"

File: solutions/gold/ac-distancequeries.mdx [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+template <class T> class BIT {
+  private:
+	int size;
+	vector<T> bit;
+	vector<T> arr;
+
+  public:
+	BIT(int size) : size(size), bit(size + 1), arr(size) {}
+
+	/** Sets the value at index ind to val. */
```
Comment (SansPapyrus683) [inline solutions/gold/ac-distancequeries.mdx:61]: no need for doc comments when in codesnip

File: solutions/gold/ac-distancequeries.mdx [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+	int lca(int u, int v) {
+		if (is_ancestor(u, v)) { return u; }
+		if (is_ancestor(v, u)) { return v; }
+		for (int i = log2dist - 1; i >= 0; i--) {
+			if (!is_ancestor(up[i][u], v)) { u = up[i][u]; }
+		}
+		return up[0][u];
+	}
+
+	void update(int i, int w) {
```
Comment (SansPapyrus683) [inline solutions/gold/ac-distancequeries.mdx:131]: doc comment for this

File: solutions/gold/ac-distancequeries.mdx [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+  private:
+	const int n, log2dist;
+	BIT<ll> bit;
+	vector<int> tin, tout;
+	vector<array<int, 3>> edges;
+	vector<vector<array<int, 2>>> adj;
+	vector<vector<int>> up;
+	int timer = 0;
+
+	void dfs(int u, int p) {
```
Comment (SansPapyrus683) [inline solutions/gold/ac-distancequeries.mdx:91]: doc comment like "does euler tour on tree to [idk]"

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
saxton hale

File: solutions/gold/ac-distancequeries.mdx (hunk 3 above)
Comment (TheGamingMousse) [reply solutions/gold/ac-distancequeries.mdx:131]: oh shoot i forgot to write doc comments LOL

**Comment by TheGamingMousse** [comment]
BTW, I used binlift to calculate LCA even though it hasn't been mentioned in the gold module yet. Is this okay, or should it be changed? (I personally think its ok b/c it's just a way to calculate the LCA, but idk)

**Comment by SansPapyrus683** [comment]
wait if we need lca, a plat topic, to solve this
maybe we should just move the problem

File: solutions/gold/ac-distancequeries.mdx [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+		up = vector<vector<int>>(log2dist, vector<int>(n));
+		dfs(0, -1);
+	}
+
+	/** @return if node u is an ancestor of node v */
+	bool is_ancestor(int u, int v) {
+		return tin[u] <= tin[v] && tout[v] <= tout[u];
+	}
+
+	/** @return the least common ancestor of nodes u and v */
```
Comment (SansPapyrus683) [inline solutions/gold/ac-distancequeries.mdx:121]: lowest common ancestor instead?

File: solutions/gold/ac-distancequeries.mdx [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+int main() {
+	int n;
+	cin >> n;
+	vector<array<int, 3>> edges(n - 1);
+	for (int i = 0; i < n - 1; i++) {
+		int u, v, w;
+		cin >> u >> v >> w;
+		u--, v--;
+		edges[i] = {u, v, w};
```
Comment (SansPapyrus683) [inline solutions/gold/ac-distancequeries.mdx:155]:
```suggestion
		edges[i] = {--u, --v, w};
```or not, commas like that just make me uncomfortable

**Review by SansPapyrus683 — APPROVED** [review]
yeah

File: solutions/gold/ac-distancequeries.mdx [hunk 7]
```diff
@@ -0,0 +1,182 @@
+---
+id: ac-distancequeries
+source: AC Beginner Contest 294
+title: Distance Queries on a Tree
+author: Justin Ji
+---
+
+<Spoiler title = "Hint">
+
+Root the tree. For an edge from $u$ -> $v$, where $u$ is
```
Comment (ryanchou-dev) [inline solutions/gold/ac-distancequeries.mdx:10]:
```suggestion
Root the tree. For an edge from $u \rightarrow v$, where $u$ is
```

File: solutions/gold/ac-distancequeries.mdx [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+of nodes?
+
+</Spoiler>
+
+<Spoiler title = "Answer to Hint 2">
+
+The distance between any pair of nodes $(u, v)$ is
+$dist(u) + dist(v) - 2 \cdot dist(lca(u, v))$,
+where $dist(x)$ equals the distance from node $x$ to the
+root.
```
Comment (ryanchou-dev) [inline solutions/gold/ac-distancequeries.mdx:28]: i think a better word is depth?

File: solutions/gold/ac-distancequeries.mdx [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+</Spoiler>
+
+<Spoiler title = "Explanation">
+
+As mentioned in the hints, we push the weights onto each node. To find the distance from any node to the root,
+we use the same approach as in [this problem](https://usaco.guide/problems/cses-1138-path-queries/solution). Each weight $w$ at each node $i$ increases the distances of every node in the subtree of $i$ by $w$, so we can use a BIT to process range additions and point queries.
+
+The solution below uses [binary jumping](https://usaco.guide/plat/binary-jump) to find the LCA between two nodes. However, the approach described in this module will also suffice.
```
Comment (ryanchou-dev) [inline solutions/gold/ac-distancequeries.mdx:37]:
```suggestion
The solution below uses [binary jumping](/plat/binary-jump) to find the LCA between two nodes. However, the approach described in this module will also suffice.
```

File: solutions/gold/ac-distancequeries.mdx [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+$dist(u) + dist(v) - 2 \cdot dist(lca(u, v))$,
+where $dist(x)$ equals the distance from node $x$ to the
+root.
+
+</Spoiler>
+
+<Spoiler title = "Explanation">
+
+As mentioned in the hints, we push the weights onto each node. To find the distance from any node to the root,
+we use the same approach as in [this problem](https://usaco.guide/problems/cses-1138-path-queries/solution). Each weight $w$ at each node $i$ increases the distances of every node in the subtree of $i$ by $w$, so we can use a BIT to process range additions and point queries.
```
Comment (ryanchou-dev) [inline solutions/gold/ac-distancequeries.mdx:35]:
```suggestion
we use the same approach as in [this problem](/problems/cses-1138-path-queries/solution). Each weight $w$ at each node $i$ increases the distances of every node in the subtree of $i$ by $w$, so we can use a BIT to process range additions and point queries.
```

**Review by ryanchou-dev — CHANGES_REQUESTED** [review]
could you also send ac screenshot? i trust you but it's just a precautionary measure

**Comment by TheGamingMousse** [comment]
> wait if we need lca, a plat topic, to solve this maybe we should just move the problem

lca is mentioned in the euler tour module tbf, i just didnt use the rmq method

File: solutions/gold/ac-distancequeries.mdx (hunk 5 above)
Comment (TheGamingMousse) [reply solutions/gold/ac-distancequeries.mdx:121]: oh shoot LOL, i was tweakin yesterday

**Comment by SansPapyrus683** [comment]
hmmmmm
would you be willing to make it rmq pretty please?

File: solutions/gold/ac-distancequeries.mdx (hunk 8 above)
Comment (TheGamingMousse) [reply solutions/gold/ac-distancequeries.mdx:28]: depth typically refers to the "level" of the node no? the graph is weighted so i think distance is more appropriate here

**Comment by TheGamingMousse** [comment]
> could you also send ac screenshot? i trust you but it's just a precautionary measure

https://atcoder.jp/contests/abc294/submissions/57018134

I am gonna update it so it uses the rmq lca method described in the module, so will send another ac submission later

File: solutions/gold/ac-distancequeries.mdx (hunk 9 above)
Comment (TheGamingMousse) [reply solutions/gold/ac-distancequeries.mdx:37]: did not know this was a thing, thanks!

File: solutions/gold/ac-distancequeries.mdx (hunk 6 above)
Comment (TheGamingMousse) [reply solutions/gold/ac-distancequeries.mdx]: changed

File: solutions/gold/ac-distancequeries.mdx (hunk 8 above)
Comment (SansPapyrus683) [reply solutions/gold/ac-distancequeries.mdx:28]: yeah i uh, agree with justin here

**Comment by TheGamingMousse** [comment]
https://atcoder.jp/contests/abc294/submissions/57018497

new submission with rmq

File: solutions/gold/ac-distancequeries.mdx (hunk 8 above)
Comment (ryanchou-dev) [reply solutions/gold/ac-distancequeries.mdx:28]: oops yeah you're right

**Review by ryanchou-dev — APPROVED** [review]

**Pull request merged**